frontends/p4/sideEffects.h

```diff
@@ -83,6 +83,46 @@ class SideEffects : public Inspector {
         expression->apply(se);
         return se.nodeWithSideEffect != nullptr;
     }
+    /// @return true if the expression may have side-effects.
+    static bool hasSideEffect(const IR::Expression* exp,
+                      ReferenceMap* refMap,
+                      TypeMap* typeMap) {
+        auto mce = exp->to<IR::MethodCallExpression>();
+        if (mce == nullptr)
+            return false;
+        // mce does not produce a side effect in the two cases:
+        //  * isValid()
+        //  * function with all in parameters
+        if (mce->method->is<IR::Member>()) {
+            auto mem = mce->method->to<IR::Member>();
+            auto basetype = typeMap->getType(mem->expr);
+            if (basetype == nullptr) {
+                basetype = mem->expr->type;
+            }
+            if (auto sc = basetype->to<IR::Type_SpecializedCanonical>())
+                basetype = sc->baseType;
+            if (basetype->is<IR::Type_HeaderUnion>()) {
+                if (mem->member == IR::Type_Header::isValid)
+                    return false;
+            } else if (basetype->is<IR::Type_Header>()) {
+                if (mem->member == IR::Type_Header::isValid)
+                    return false;
+            }
+        } else if (mce->method->is<IR::PathExpression>()) {
+            auto pe = mce->method->to<IR::PathExpression>();
+            auto decl = refMap->getDeclaration(pe->path, true);
+            if (decl->is<IR::Function>()) {
+                auto mi = MethodInstance::resolve(mce, refMap, typeMap);
```

Comment:
Why not use Resolve first and the look at the result to categorize the function? 
It does most of the work for you. It even has a case for built-in methods.